### DISCUSSION TITLE
Is there a way to make a type serializable for xunit without implementing the IXUnitSerializable interface

### DISCUSSION THREAD

**Discussion by sensslen** [xunit/xunit#2956] [Question] [answered]
I'm currently facing situations where I have parametric tests that use a data type I do not control. I would be able to serialize/deserialize easily but I cannot implement the IXUnitSerializable because it's a sealed type....

**Comment by bradwilson** [comment]
Unfortunately, that's the only way to do it, unless you can somehow split your data into already supported types and reassemble it into the data type you don't control (typically done by embedded data types you don't control into ones you do, and then implementing the serialization on the outer type).

**Comment by sensslen** [comment]
@bradwilson thanks for the response. That's a bummer but I get it.

**Comment by markusschaber** [comment]
@bradwilson What do you think about providing external serializer implementations? They could be brought in with some attribute, either at the TheoryData member, or class or assembly level. Similar in spirit to the [`\[JsonConverter\]`](https://learn.microsoft.com/en-us/dotnet/api/system.text.json.serialization.jsonconverterattribute?view=net-8.0) Attribute.

As far as I can see, actually using the convention defined by the [`JsonConverter`](https://learn.microsoft.com/en-us/dotnet/api/system.text.json.serialization.jsonconverter?view=net-8.0) abstract class from System.Text.Json would be very nice, allowing to reuse existing serialization infrastructure including custom serializers already written for exactly those data types used by the SUT, as quite some of the datatypes involved in the SUT are already JSON serializable, for one or another reason.

Heck, not even standard types like Uri or Guid are considered serializable by xunit, and they both are commonly used and have straight forward, lossless string representations, and fit nicely with JSON.

According to https://xunit.net/xunit.analyzers/rules/xUnit1045 the Visual Studio Test Explorer (which I don't even use) is the reason for this restriction, which I can understand technically. But restricting the basic types to not include Uri or GUID seems a bit arbitrary, and restricting the extension point to an interface to be implemented by the test data also seems limiting. And the ability to reuse existing serialization implementations would be very useful.

**Reply by bradwilson** [reply]
> actually using the convention defined by the [JsonConverter](https://learn.microsoft.com/en-us/dotnet/api/system.text.json.serialization.jsonconverter?view=net-8.0) abstract class from System.Text.Json would be very nice

We don't take dependencies on things that don't ship in the runtime, because our dependency version trumps yours.

Plus, if we're talking about v2, `System.Text.Json` isn't available for our minimum target frameworks (`net452` and `netcoreapp1.0`).

> restricting the basic types to not include Uri or GUID seems a bit arbitrary

It's 100% arbitrary in the sense that "we support what we support" (i.e., we have to hand craft support). We have added new types over time, so open an issue to add support for `Uri` and `Guid`.

> the ability to reuse existing serialization implementations would be very useful

Again, if we're talking about v2, there are none in `netstandard1.1`.

**Reply by markusschaber** [reply]
Ok, I understand. Then, the alternative could be an `IXunitSerializer` interface, which could look like this:
```
public interface IXunitSerializer<TData> {
    string Serialize(TData data);
    TData Deserialize(string serialized);
}
```
It's trivial to implement that interface for System.Text.Json or NewtonSoft Json - trivial enough that an example in the documentation should be enough (and AFAICS also trivial enough that it's not copyrightable, thus automatically public domain):
```
public class XunitJsonSerializer<TData> : IXunitSerializer<TData>
{
    public string Serialize(TData data) => JsonSerializer.Serialize(data);
        
    public TData Deserialize(string serialized) => JsonSerializer.Deserialize<TData>(serialized)!;
}

public class XUnitNewtonSoftSerializer<TData> : IXunitSerializer<TData>
{
    public string Serialize(TData data) => Newtonsoft.Json.JsonConvert.SerializeObject(data);
    public TData Deserialize(string serialized) => Newtonsoft.Json.JsonConvert.DeserializeObject<TData>(serialized)!;
}
```
Then, those serializers could be declared at either the TheoryData member, the class, or the assembly level via an attribute:
`[assembly: XUnitSerializers(typeof( XUnitJsonSerializer<MyDataStructure>), typeof(XUnitNewtonSoftSerializer<OtherDataStructure>), typeof(SomeTotallyCustomConverter))]`

Alternatively, the `IXunitSerializer` interface could be defined on `object` and `IXunitSerializationInfo`, to keep things close to what you already have.

**Reply by markusschaber** [reply]
Just as a clarification: While adding `Guid` and `Uri` to the set of supported intrinsics solves quite some of our theories, we have others where we use Classes or Interfaces predefined by external code (our business logic, or external libraries). We could wrap each of those in a `IXunitSerializable` wrapper class, but that still seems a bit clumsy compared to providing an external serializer implementation.

I even just realized that all the built-in support (including `IXUnitSerializable`) could be implemented via `IXunitSerializer<T>` implementations which are just registered by default, removing the need for special casing altogether.

**Reply by bradwilson** [reply]
Registration of an external serializer is an interesting idea. I just shifted runner reporters over to a similar model where all [the reporters are linked into the test assembly and registered via attribute](https://github.com/xunit/xunit/issues/1874). This fixed a fundamentally broken model for runner reporters in v2; trying to do registered serializers in v2 would've been similarly broken because the code would need to live in both sides of the system (in the runners, and in the test assembly), which is fundamentally altered by the fact that the runner now lives inside the test assembly.

So... maybe this model would work here as well. I'll try it out and see.

**Reply by bradwilson** [reply]
FWIW, I'm not a fan of the generic here, for a few reasons:

- Generics aren't free, they add runtime cost
- It adds reflection logic to extract the type at runtime during registration
- It forces a serializer to either be 100% generic, or 100% specific (which would cause some limited serializers to have to implement the interface multiple times)

I'm almost 100% sure I'd end up with this instead:

```csharp
public interface IXunitSerializer {
    string Serialize(object data);
    object Deserialize(Type type, string serialized);
}
```

And then something along the lines of a registration which allows you to specify supported types in the attribute rather than embedded into the class, like:

```csharp
[assembly: XunitSerializer(typeof(XunitJsonSerializer), typeof(MyDataStructure), typeof(MyOtherDataStructure))]
```

It's too early to tell, though, until I start trying and validating designs.

**Reply by markusschaber** [reply]
Yes, generics may come with their downsides, especially when you need to generate code invoking them.

The `object` way together with explicitly declaring the types via attributes looks good to me.

The matching of supported types should allow polymorphism, e. G. via [`IsAssignableFrom`](https://learn.microsoft.com/en-us/dotnet/api/system.type.isassignablefrom?view=net-8.0) or [`IsAssignableTo`](https://learn.microsoft.com/en-us/dotnet/api/system.type.isassignableto?view=net-8.0), instead of an exact type match. This allows to support polymorphic serialization, like using the [`\[JsonDerivedType\]`](https://docs.microsoft.com/en-us/dotnet/api/System.Text.Json.Serialization.JsonDerivedTypeAttribute?view=net-8.0) attribute.

**Reply by bradwilson** [reply]
Yes, this is how it has been implemented, and both `IXunitSerializable` and enum values are done via `IXunitSerializer` now. Exact type match takes preference over `IsAssignableFrom` match: https://github.com/xunit/xunit/blob/685423ea7942d71ef4f6781ff3fc216c0257e64c/src/xunit.v3.common/Serialization/SerializationHelper.cs#L463-L479

**Reply by markusschaber** [reply]
Thank you very much. I'm looking forward to it.

**Comment by bradwilson** [comment, marked as answer]
`IXunitSerializer` has been added in #3040.

Available in v3 `0.5.0-pre.20` https://xunit.net/docs/using-ci-builds

**Comment by markusschaber** [comment]
Just another remark which might be useful for others:
We had a theory where we had non-serializable data. This should be rare, but it can happen if the data involve Delegates with closures, or external Resources (Streams) or whatever.
As a workaround, we derived strings identifying the test data, and put them in a dictionary. The TheoryData now contains the string representations, and the first line of the test case does the lookup in the dictionary. This way, one can work around data types which cannot (or should not) be serialized.

**Comment by ViRuSTriNiTy** [comment]
Where can I find examples on how to use the new `IXunitSerializer` feature? The documentation at https://xunit.net/docs/getting-started/v3/custom-serialization#serialization-support-in-v3 is just too vague for me.

**Reply by bradwilson** [reply]
There are a few implemented inside xUnit.net itself. The simplest is probably the thing that supports `IXunitSerializable` values, since it just turns around and calls a different serializer. The enum one is probably the 2nd simplest, since it relies on `Convert.ToInt64` and `Convert.ToUInt64` to serialize, and `Enum.Parse` to deserialize.

* Enum values: [EnumSerializer](https://github.com/xunit/xunit/blob/f73db5fbfeb1ba804af03def266b1d0486205409/src/xunit.v3.common/Serialization/Serializers/EnumSerializer.cs)
* `IFormattable`/`IParsable<T>` values: [FormattableAndParsableSerializer](https://github.com/xunit/xunit/blob/f73db5fbfeb1ba804af03def266b1d0486205409/src/xunit.v3.common/Serialization/Serializers/FormattableAndParsableSerializer.cs)
* Tuple values: [TupleSerializer](https://github.com/xunit/xunit/blob/f73db5fbfeb1ba804af03def266b1d0486205409/src/xunit.v3.common/Serialization/Serializers/TupleSerializer.cs)
* `IXunitSerializable` values: [XunitSerializableSerializer](https://github.com/xunit/xunit/blob/f73db5fbfeb1ba804af03def266b1d0486205409/src/xunit.v3.common/Serialization/Serializers/XunitSerializableSerializer.cs)
